engine/classes/Elgg/Notifications/NotificationEventHandler.php

```diff
@@ -0,0 +1,356 @@
+	 * @return bool
+	 */
+	protected function sendNotification($guid, $method, array $params = []) {
+		if (!_elgg_services()->hooks->hasHandler('send', "notification:{$method}")) {
+			// no way to deliver given the current method, so quitting early
+			return false;
+		}
+
+		$actor = $this->event->getActor();
+		$object = $this->event->getObject();
+
+		if ($this->event instanceof InstantNotificationEvent) {
+			$recipient = _elgg_services()->entityTable->get($guid);
+			/* @var \ElggEntity $recipient */
+			$subject = elgg_extract('subject', $params, '');
+			$body = elgg_extract('body', $params, '');
+			$summary = elgg_extract('summary', $params, '');
+		} else {
+			$recipient = _elgg_services()->entityTable->get($guid, 'user');
+			/* @var \ElggUser $recipient */
+			if (!$recipient || $recipient->isBanned()) {
+				return false;
+			}
+		
+			if ($recipient->guid === $this->event->getActorGUID()) {
+				// Content creators should not be receiving subscription
+				// notifications about their own content
+				return false;
+			}
+			
+			if (!$actor || !$object) {
+				return false;
+			}
+
+			if ($object instanceof ElggEntity && !_elgg_services()->accessCollections->hasAccessToEntity($object, $recipient)) {
+				// Recipient does not have access to the notification object
+				// The access level may have changed since the event was enqueued
+				return false;
+			}
+
+			$subject = $this->getNotificationSubject($recipient);
+			$body = $this->getNotificationBody($recipient);
+			$summary = '';
+		}
+
+		$language = $recipient->getLanguage();
+		$params['event'] = $this->event;
+		$params['method'] = $method;
+		$params['sender'] = $actor;
+		$params['recipient'] = $recipient;
+		$params['language'] = $language;
+		$params['object'] = $object;
+		$params['action'] = $this->event->getAction();
+		$params['add_salutation'] = elgg_extract('add_salutation', $params, true);
+
+		$notification = new Notification($actor, $recipient, $language, $subject, $body, $summary, $params);
+
+		$notification = _elgg_services()->hooks->trigger('prepare', 'notification', $params, $notification);
+		if (!$notification instanceof Notification) {
+			throw new RuntimeException("'prepare','notification' hook must return an instance of " . Notification::class);
+		}
+
+		$type = 'notification:' . $this->event->getDescription();
+		$notification = _elgg_services()->hooks->trigger('prepare', $type, $params, $notification);
+		if (!$notification instanceof Notification) {
+			throw new RuntimeException("'prepare','{$type}' hook must return an instance of " . Notification::class);
+		}
+
+		if (elgg_extract('add_salutation', $notification->params) === true) {
+			$viewtype = elgg_view_exists('notifications/body') ? '' : 'default';
+			$notification->body = _elgg_view_under_viewtype('notifications/body', ['notification' => $notification], $viewtype);
+		}
+		
+		$notification = _elgg_services()->hooks->trigger('format', "notification:$method", [], $notification);
```

Comment:
`notification:{$method}` for readability

Reply:
done